### DISCUSSION TITLE
[HW Accel Support]:  for AMD system running on Debian and Docker Compose

### DISCUSSION THREAD

**Discussion by Grover-R2** [blakeblackshear/frigate#10407] [Hardware Acceleration Support]
### Describe the problem you are having

cameras are crashing whenever I try to start Frigate with hardware acceleration.  This is running on an a headless Debian 12 machine with Docker Compose.  Processor is: AMD A8-5557M APU with Radeon(tm) HD Graphics.  I've re-installed the MESA drivers to make sure they are running plus put LIBVA_DRIVER_NAME=radeonsi in the docker compose file along with the system environment but it always fails when I uncomment the hwaccel_args: preset-vaapi in the frigate config file. Looking at the log files perhaps I have it in the wrong place in the config file but I've move,  but I moved it everywhere I think possible.  Thanks in advance for any and all help.  this has been driving crazy.

### Version

0.13.2-6476F8A

### Frigate config file

```yaml
## yaml-language-server: $schema=http://x.y.z.98:5000/api/config/schema.json
#
ffmpeg:
  hwaccel_args: preset-vaapi
mqtt:
 enabled: False
#detectors:
# coral:
#    type: edgetpu
#    device: pci:0
# cpu01:
#    type: cpu
objects:
  track:
    - person
    - car
    - cat
    - dog  
record:
  enabled: True
  retain:
    days: 7
    mode: motion
  events:
    retain:
      default: 30
      mode: motion
snapshots:
  enabled: True
  retain:
    default: 30

cameras:
  Furnace: 
    enabled: True
    ffmpeg:
      inputs:
        - path: rtsp://*:*@x.y.z.47:554/cam/realmonitor?channel=1&subtype=0&unicast=true&proto=Onvif
          roles:
            - record
        - path: rtsp://*:*@x.y.z.47:554/cam/realmonitor?channel=1&subtype=1&unicast=true&proto=Onvif
          roles:
            - detect
  Drive-Way:
    enabled: true
    ffmpeg:
      inputs:
        - path: rtsp://*:*@x.y.z.50:554/Streaming/Channels/1?transportmode=unicast&profile=Profile_1
          roles:
            - record
        - path: rtsp://*:*@x.y.z.50:554/Streaming/Channels/2?transportmode=unicast&profile=Profile_2
          roles: 
            - detect 
  Lab-Revotech:          
    enabled: true
    ffmpeg:
      inputs:
        - path: rtsp://*:*@x.y.z.46:554/profile0
          roles:
            - record
        - path: rtsp://*:*@x.y.z.46:554/profile1
          roles: 
            - detect
  Front-Door: 
    enabled: True
    ffmpeg:
      inputs:
        - path: rtsp://*:*@x.y.z.44:554/cam/realmonitor?channel=1&subtype=0&unicast=true&proto=Onvif
          roles:
            - record
        - path: rtsp://*:*@x.y.z.44:554/cam/realmonitor?channel=1&subtype=1&unicast=true&proto=Onvif
          roles:
            - detect
```

### docker-compose file or Docker CLI command

```yaml
# I also inculde LIBVA_DRIVER_NAME in the system environment as well.
version: "3.9"
#environment:
#  - LIBVA_DRIVER_NAME=radeonsi
services:
  frigate:
    container_name: frigate
    privileged: true
    restart: unless-stopped
    image: ghcr.io/blakeblackshear/frigate:stable
    shm_size: "128mb"
    devices:
      - /dev/dri/renderD128
    volumes:
      - ./frigate/config:/config
      - ./frigate/storage:/media/frigate
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "5000:5000"
      - "8554:8554" # RTSP feeds
    environment:
      - LIBVA_DRIVER_NAME=radeonsi
```

### Relevant log output

```shell
2024-03-12 19:29:17.406856611  [2024-03-12 19:29:17] ffmpeg.Drive-Way.detect        ERROR   : Option hwaccel (use HW accelerated decoding) cannot be applied to output url vaapi -- you are trying to apply an input option to an output file or vice versa. Move this option before the file it belongs to.
2024-03-12 19:29:17.407122098  [2024-03-12 19:29:17] ffmpeg.Drive-Way.detect        ERROR   : Error parsing options for output file vaapi.
2024-03-12 19:29:17.407129481  [2024-03-12 19:29:17] ffmpeg.Drive-Way.detect        ERROR   : Error opening output files: Invalid argument
2024-03-12 19:29:17.419253002  [2024-03-12 19:29:17] ffmpeg.Furnace.record          ERROR   : Option hwaccel (use HW accelerated decoding) cannot be applied to output url vaapi -- you are trying to apply an input option to an output file or vice versa. Move this option before the file it belongs to.
2024-03-12 19:29:17.434682457  [2024-03-12 19:29:17] ffmpeg.Furnace.record          ERROR   : Error parsing options for output file vaapi.
2024-03-12 19:29:17.434691590  [2024-03-12 19:29:17] ffmpeg.Furnace.record          ERROR   : Error opening output files: Invalid argument
2024-03-12 19:29:17.434695941  [2024-03-12 19:29:17] watchdog.Furnace               INFO    : Terminating the existing ffmpeg process...
2024-03-12 19:29:17.434701396  [2024-03-12 19:29:17] watchdog.Furnace               INFO    : Waiting for ffmpeg to exit gracefully...
2024-03-12 19:29:17.439274845  [2024-03-12 19:29:17] ffmpeg.Drive-Way.record        ERROR   : Option hwaccel (use HW accelerated decoding) cannot be applied to output url vaapi -- you are trying to apply an input option to an output file or vice versa. Move this option before the file it belongs to.
2024-03-12 19:29:17.456372751  [2024-03-12 19:29:17] ffmpeg.Drive-Way.record        ERROR   : Error parsing options for output file vaapi.
2024-03-12 19:29:17.456381416  [2024-03-12 19:29:17] ffmpeg.Drive-Way.record        ERROR   : Error opening output files: Invalid argument
2024-03-12 19:29:17.456385009  [2024-03-12 19:29:17] watchdog.Drive-Way             INFO    : Terminating the existing ffmpeg process...
2024-03-12 19:29:17.456388020  [2024-03-12 19:29:17] watchdog.Drive-Way             INFO    : Waiting for ffmpeg to exit gracefully...
2024-03-12 19:29:17.461478317  [2024-03-12 19:29:17] watchdog.Lab-Revotech          ERROR   : Ffmpeg process
```

### FFprobe output from your camera

```shell
#
# Revotech Camera:
# FFPROBE OUTPUT
[{"return_code":0,"stderr":"","stdout":{"programs":[],"streams":[{"avg_frame_rate":"8/1","codec_long_name":"H.264/AVC/MPEG-4AVC/MPEG-4part10","height":1080,"width":1920},{"avg_frame_rate":"0/0","bit_rate":"64000","codec_long_name":"PCMmu-law/G.711mu-law"}]}},{"return_code":0,"stderr":"","stdout":{"programs":[],"streams":[{"avg_frame_rate":"5/1","codec_long_name":"H.264/AVC/MPEG-4AVC/MPEG-4part10","height":480,"width":720},{"avg_frame_rate":"0/0","bit_rate":"64000","codec_long_name":"PCMmu-law/G.711mu-law"}]}}]
#
# Amercrest Camera
#
[{"return_code":0,"stderr":"","stdout":{"programs":[],"streams":[{"avg_frame_rate":"5/1","codec_long_name":"H.264/AVC/MPEG-4AVC/MPEG-4part10","height":360,"width":640},{"avg_frame_rate":"0/0","bit_rate":"64000","codec_long_name":"PCMmu-law/G.711mu-law"}]}},{"return_code":0,"stderr":"","stdout":{"programs":[],"streams":[{"avg_frame_rate":"5/1","codec_long_name":"H.264/AVC/MPEG-4AVC/MPEG-4part10","height":360,"width":640},{"avg_frame_rate":"0/0","bit_rate":"64000","codec_long_name":"PCMmu-law/G.711mu-law"}]}}]
#
# Trendnet
#
[{"return_code":0,"stderr":"","stdout":{"programs":[],"streams":[{"avg_frame_rate":"8/1","codec_long_name":"H.264/AVC/MPEG-4AVC/MPEG-4part10","height":720,"width":1280}]}},{"return_code":0,"stderr":"","stdout":{"programs":[],"streams":[{"avg_frame_rate":"0/0","codec_long_name":"MotionJPEG","display_aspect_ratio":"22:15","height":240,"width":352}]}}]
```

### Operating system

Debian

### Install method

Docker Compose

### Network connection

Wired

### Camera make and model

Amcrest, trendnet TV-IP310PI, & Revotech FH8852

**Comment by NickM-27** [comment]
what is vainfo output (from the frigate system page)

**Reply by Grover-R2** [reply]
Return Code: 1
Process Error:
error: XDG_RUNTIME_DIR is invalid or not set in the environment.
error: can't connect to X server! 
error: failed to initialize display

this issue unfortunately is at or beyond my understanding of linux and docker I'm afraid.  I see it exists and set in my system environment to:
XDG_RUNTIME_DIR=/run/user/1000

and this directory actually exists too.

**Reply by NickM-27** [reply]
there's no reason to worry about this. The problem is simply that the GPU is not visible in the frigate container. Do you see /dev/dri/renderD128 in the frigate container (by getting a shell and running ls /dev/dri)

**Reply by Grover-R2** [reply]
So I opened a shell to the frigate container. 
docker exec -it frigate /bin/bash
within the container /dev I don't see  /dri at all.  outside docker its there along with renderD128.

**Reply by NickM-27** [reply]
What happens if you run vainfo on the host

**Reply by Grover-R2** [reply]
I get the same answer as from vainfo from the system tab within frigate
root@ce3019b54316:/# vainfo
error: XDG_RUNTIME_DIR is invalid or not set in the environment.
error: can't connect to X server!
error: failed to initialize display

**Reply by NickM-27** [reply]
Send be driver isn't installed correctly on the host

**Reply by Grover-R2** [reply]
Just to follow up I built several other Frigate installations all on circa 2016-2018 hardware I had around the shop.  Couldn't get hwaccel to work on any of them intel and AMD.  Just received a Beelink e12 and hwaccel basically worked out of the box!  now to get Coral TPU working....(hardware not received yet). Thanks for your help @NickM-27.

**Comment by PablexXXXX** [comment]
I was able to make it work with this line

    environment:
       LIBVA_DRIVER_NAME: radeonsi

Take a look at your docker compose file and see the differences

**Comment by trythatagain** [comment]
I see in your compose file you have
`    devices:
      - /dev/dri/renderD128
 `

I have
`      - /dev/dri/renderD128:/dev/dri/renderD128`
`      - /dev/dri/card0:/dev/dri/card0`

Not sure if you need the card0 line, but this is what I run for my AMD 5600G system.

**Comment by g13092** [comment]
@Grover-R2  Did you ever find a solution?  I just found your thread looking for answers to the same problem on my junk pc...

**Reply by Grover-R2** [reply]
I unfortunately did not get around to trying what @trythatagain again suggested.  And thanks for all the suggestions.  I kinda took these easy way out and bought the exact same make and model Beelink that is mentioned in the wiki.  And wow it works like a charm.   Got it from Amazon, had it up in running with HW AACCEL and Coral TPU in like 45 min.  this was made a little easier as I already had a 90% config.  have it up and running with 8 cameras of various makes and models.

**Reply by g13092** [reply]
@Grover-R2 thanks for the update!  I'm thinking of trying out another PC I have lying around or look for a cheap 'up to date' video card I could slap in.  

I tried the       - /dev/dri/renderD128:/dev/dri/renderD128 but not the card0 and that didnt help.   I think it tries to use the integrated hardware but that somehow that hardware is not up to the task.
